### DISCUSSION TITLE
VMQ diversity module with Oracle DB support

### DISCUSSION THREAD

**Discussion by vshirol** [vernemq/vernemq#2198] [General]
Hi All,

We are using VMQ diversity module for lua plugin in VerneMQ along with redis. This works fine, now we are planning to make use OracleDB and keep auth information in Oracle DB. We don't see any configuration properties inside vernemq.conf related oracle DB, although there are properties for postgres/mysql.

So can we make use of mysql related properties for connecting Oracle DB? Is this going to work or any other way?

Thanks & Regards,
Vasant

**Comment by ioolkos** [comment]
Hi @vshirol,

thanks for your question. The Lua scripts in `vmq_diversity` are database specific because they are based on DB-specific Erlang drivers.
So the MySQL driver cannot talk to Oracle. Well, I haven't tried... but it'd be too good to be true :)

The options here are either developing an Oracle-specific `vmq_diversity` plugin (modelled after the existing SQL plugins), or a more generic ODBC plugin, using the ODBC layer that's part of Erlang. (it can talk to Oracle). It doesn't sound too difficult to do, there's going to bit a bit of complexity coming from integration into Luerl.

Oh, speaking of Lua, your third option, of course, is developing a direct Erlang authentication plugin without using the `vmq_diversity` Lua stuff.
If you or anyone else want to have a go at this, let us know.

**Reply by vshirol** [reply]
Thanks @ioolkos for your quick response and suggestions.

**Comment by ioolkos** [comment]
There's an initial attempt to add ODBC to vmq_diversity here: https://github.com/vernemq/vernemq/pull/2298
I haven't tested this with Oracle, though. (with Postgres ODBC driver it works)

**Reply by mths1** [reply]
@ioolkos : How good (or bad) is the ODBC implementation in Erlang? Coming from a Java backround where JDBC is the standard, I am wondering, if it makes sense to really keep all those custom erlang-based DB drivers which cause effort for maintainance and where we have quite some dependencies to third-parties.

**Reply by ioolkos** [reply]
@mths1 it's basic and only tested for MSSql and Postgres. It could spare us the Postgres and MySQL drivers only, at the price of requiring an ODBC driver. Plus it's a C Nif.
So, I'd see it as a supplement, not a replacement.
